### PR TITLE
feat: add gcp iap auth provider

### PR DESCRIPTION
**Pull request by FabianHippmann**
## Hey, I just made a Pull Request!
This PR adds the Google Cloud Platform Identity Aware proxy Authentication provider. 

     
Discussed in - https://github.com/backstage/backstage/issues/5564

Tests need to be fixed & -> https://github.com/backstage/backstage/issues/6245 could be adapted

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by weyert** [comment]
@FabianHippmann Are you planning completing this PR? As I would love to be able to use GCP IAP as authentication provider

File: plugins/auth-backend/src/providers/gcp-iap/provider.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    logger: Logger,
+    catalogClient: CatalogApi,
+    options: GcpIAPProviderOptions,
+  ) {
+    this.logger = logger;
+    this.catalogClient = catalogClient;
+    this.options = options;
+  }
+  frameHandler(): Promise<void> {
+    return Promise.resolve(undefined);
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/gcp-iap/provider.ts:51]: I think throwing some `NotImplementedError`s in these make sense

File: plugins/auth-backend/src/providers/gcp-iap/provider.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+  async refresh(req: express.Request, res: express.Response): Promise<void> {
+    const expectedAudience = this.options.audience;
+    const jwtToken = req.header(IAP_JWT_HEADER);
+    if (jwtToken === undefined) {
+      res.status(401);
+      res.end();
+      return;
+    }
+    const oAuth2Client = new OAuth2Client();
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/gcp-iap/provider.ts:62]: I'm assuming this contains state that we want to keep around, cached public keys and so on. So prolly better in keep this in a member

File: plugins/auth-backend/src/providers/gcp-iap/provider.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    try {
+      const user = await verify();
+      if (user === undefined) {
+        this.logger.error('gcp iap proxy user returned undefined');
+        res.status(401);
+        res.end();
+        return;
+      }
+      const resolvedEntity = await this.options.identityResolutionCallback(
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/gcp-iap/provider.ts:82]: Let's use the pattern described [here](https://github.com/backstage/backstage/issues/6245) instead, with the `fullProfile` being the `user` here. It also opens up for some custom population of the profile, which we at least end up having to do internally to fetch profile pictures.

File: plugins/auth-backend/src/providers/gcp-iap/provider.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        ['https://cloud.google.com/iap'],
+      );
+      return ticket.getPayload();
+    };
+
+    try {
+      const user = await verify();
+      if (user === undefined) {
+        this.logger.error('gcp iap proxy user returned undefined');
+        res.status(401);
```
Comment (Rugvip) [inline plugins/auth-backend/src/providers/gcp-iap/provider.ts:78]: Let's just throw errors instead so that we leave the error handling more flexible

**Review by Rugvip — COMMENTED** [review]
Thank you! 👍 

Looks good, a few nits and let's move to using the pattern that's being established for all auth providers around custom user resolution logic.

Also thinking it's time to provide a built-in component that handles this sign-in flow in the frontend, but that's a separate PR

**Comment by Rugvip** [comment]
Hi @FabianHippmann, we can help out get this PR in shape, but would need the DCO check to be fixed in order to be able to work on top of it. Are you around to get that piece fixed? :grin: Unfortunately we'll have to close this otherwise

**Comment by adamdmharvey** [comment]
Brief ping on this one.  @FabianHippmann , do you think you might be able to check the DCO via tips here for your existing commits to this excellent addition?

 https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin  

Any further plans to move forward on this? Just helping align.  Thanks!

**Comment by FabianHippmann** [comment]
@adamdmharvey - thank you for the reminder - I signed the commits 

I'll get an overview overview of the changes & push this week to get a mergeable commit!

**Comment by danieljimeneznz** [comment]
Hi @FabianHippmann @Rugvip I'm keen to make use of this and help get this PR across the line - is this something that you think would still be beneficial to the backstage community?

**Comment by freben** [comment]
@danieljimeneznz it's actually in! We took it upon ourselves to finish it off some time ago in a separate PR which replaces this one.

https://github.com/backstage/backstage/pull/8589

**Comment by danieljimeneznz** [comment]
@freben awesome! - I couldn't find any documentation about it on the backstage docs page?

**Comment by freben** [comment]
Ah, hm it doesn't seem to be linked from the sidebar, must have missed that. But it's there! 

https://github.com/backstage/backstage/blob/master/docs/auth/google/gcp-iap-auth.md

**Pull request closed without merging**